File: imagepipeline-test/src/main/java/com/facebook/imagepipeline/testing/MockBitmapFactory.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.testing;

import static org.mockito.Mockito.*;

import android.graphics.Bitmap;
import com.facebook.common.internal.Preconditions;
import com.facebook.imageutils.BitmapUtil;

/**
 * Helper class for creating bitmap mocks in tests.
 */
public class MockBitmapFactory {
  public static int DEFAULT_BITMAP_WIDTH = 3;
  public static int DEFAULT_BITMAP_HEIGHT = 4;
  public static int DEFAULT_BITMAP_PIXELS = DEFAULT_BITMAP_WIDTH * DEFAULT_BITMAP_HEIGHT;
  public static int DEFAULT_BITMAP_SIZE = bitmapSize(
      DEFAULT_BITMAP_WIDTH,
      DEFAULT_BITMAP_HEIGHT,
      Bitmap.Config.ARGB_8888);

  public static Bitmap create() {
    return create(DEFAULT_BITMAP_WIDTH, DEFAULT_BITMAP_HEIGHT, Bitmap.Config.ARGB_8888);
  }

  public static Bitmap createForSize(int size, Bitmap.Config config) {
    Preconditions.checkArgument(size % BitmapUtil.getPixelSizeForBitmapConfig(config) == 0);
    return create(1, size / BitmapUtil.getPixelSizeForBitmapConfig(config), config);
  }

  public static Bitmap create(int width, int height, Bitmap.Config config) {
    Preconditions.checkArgument(width > 0);
    Preconditions.checkArgument(height > 0);
    Preconditions.checkNotNull(config);
    Bitmap bitmap = mock(Bitmap.class);
    when(bitmap.getWidth()).thenReturn(width);
    when(bitmap.getHeight()).thenReturn(height);
    when(bitmap.getConfig()).thenReturn(config);
    when(bitmap.isMutable()).thenReturn(true);
    when(bitmap.getRowBytes()).thenReturn(width * BitmapUtil.getPixelSizeForBitmapConfig(config));
    when(bitmap.getByteCount()).thenReturn(bitmapSize(width, height, config));
    return bitmap;
  }

  public static int bitmapSize(int width, int height, Bitmap.Config config) {
    return BitmapUtil.getSizeInByteForBitmap(width, height, config);
  }
}

File: imagepipeline-test/src/main/java/com/facebook/imagepipeline/testing/MockBitmapFactory.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.testing

import android.graphics.Bitmap
import com.facebook.imageutils.BitmapUtil
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

/** Helper class for creating bitmap mocks in tests. */
object MockBitmapFactory {

  const val DEFAULT_BITMAP_WIDTH: Int = 3
  const val DEFAULT_BITMAP_HEIGHT: Int = 4
  const val DEFAULT_BITMAP_PIXELS: Int = DEFAULT_BITMAP_WIDTH * DEFAULT_BITMAP_HEIGHT
  @get:JvmStatic
  val DEFAULT_BITMAP_SIZE: Int =
      bitmapSize(DEFAULT_BITMAP_WIDTH, DEFAULT_BITMAP_HEIGHT, Bitmap.Config.ARGB_8888)

  @JvmStatic
  fun createForSize(size: Int, config: Bitmap.Config): Bitmap {
    require(size % BitmapUtil.getPixelSizeForBitmapConfig(config) == 0)
    return create(1, size / BitmapUtil.getPixelSizeForBitmapConfig(config), config)
  }

  @JvmStatic
  @JvmOverloads
  fun create(
      width: Int = DEFAULT_BITMAP_WIDTH,
      height: Int = DEFAULT_BITMAP_HEIGHT,
      config: Bitmap.Config = Bitmap.Config.ARGB_8888,
  ): Bitmap {
    require(width > 0)
    require(height > 0)
    checkNotNull(config)
    val bitmap = mock<Bitmap>()
    whenever(bitmap.width).thenReturn(width)
    whenever(bitmap.height).thenReturn(height)
    whenever(bitmap.config).thenReturn(config)
    whenever(bitmap.isMutable).thenReturn(true)
    whenever(bitmap.rowBytes).thenReturn(width * BitmapUtil.getPixelSizeForBitmapConfig(config))
    whenever(bitmap.byteCount).thenReturn(bitmapSize(width, height, config))
    return bitmap
  }

  @JvmStatic
  fun bitmapSize(width: Int, height: Int, config: Bitmap.Config): Int =
      BitmapUtil.getSizeInByteForBitmap(width, height, config)
}
